server/project_member.go

```diff
@@ -2,18 +2,186 @@ package server
 	"github.com/google/jsonapi"
 	"github.com/labstack/echo/v4"
 	"go.uber.org/zap"
 )
 
 func (s *Server) registerProjectMemberRoutes(g *echo.Group) {
+	// for now we only support sync project member from privately deployed GitLab
+	g.POST("/project/:projectID/syncmember", func(c echo.Context) error {
+		ctx := context.Background()
+		projectID, err := strconv.Atoi(c.Param("projectID"))
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Project ID is not a number: %s", c.Param("projectID"))).SetInternal(err)
+		}
+
+		projectFind := &api.ProjectFind{ID: &projectID}
+		project, err := s.ProjectService.FindProject(ctx, projectFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Project not found: %s", c.Param("projectID"))).SetInternal(err)
+		}
+		if project.WorkflowType != api.VCSWorkflow {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid workflow type: %s, need %s to enable this function", project.WorkflowType, api.VCSWorkflow))
+		}
+
+		repoFind := &api.RepositoryFind{ProjectID: &projectID}
+		repo, err := s.RepositoryService.FindRepository(ctx, repoFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch relevant VCS repo, Project ID: %s", c.Param("projectID"))).SetInternal(err)
+		}
+
+		vcsFind := &api.VCSFind{
+			ID: &repo.VCSID,
+		}
+		vcs, err := s.VCSService.FindVCS(ctx, vcsFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find VCS for sync project member: %d", repo.VCSID)).SetInternal(err)
+		}
+		if vcs == nil {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("VCS ID not found: %d", repo.VCSID))
+		}
+
+		vcsProjectMemberList, err := vcsPlugin.Get(vcs.Type, vcsPlugin.ProviderConfig{Logger: s.l}).FetchRepositoryActiveMemberList(ctx,
+			common.OauthContext{
+				ClientID:     vcs.ApplicationID,
+				ClientSecret: vcs.Secret,
+				AccessToken:  repo.AccessToken,
+				RefreshToken: repo.RefreshToken,
+				Refresher:    s.refreshToken(ctx, repo.ID),
+			},
+			vcs.InstanceURL,
+			repo.ExternalID,
+		)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch repository member from VCS, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+		}
+
+		findProjectMember := &api.ProjectMemberFind{ProjectID: &projectID}
+		bytebaseProjectMemberList, err := s.ProjectMemberService.FindProjectMemberList(ctx, findProjectMember)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch repository member in bytebase: %d", projectID)).SetInternal(err)
+		}
+
+		// drop all existing members
+		{
+			for _, member := range bytebaseProjectMemberList {
+				memberDelete := &api.ProjectMemberDelete{
+					ID:        member.ID,
+					DeleterID: c.Get(getPrincipalIDContextKey()).(int),
```

Comment:
We may want a same database transaction to update members (delete+add/etc) just in case something fail in the middle causing a state not recoverable.